Add caps lock indicator

codokie opened the pull request:
I made some changes to the different shift states icons.
(note: the color of the theme is used in shifted/caps lock states, below I used the Brown color but in practice it can be whatever color is chosen in the theme settings)
**unshifted state:**
[material]
![unshifted_material](https://github.com/Helium314/HeliBoard/assets/151087174/20540aba-1943-4e67-84be-72c812371e1e)

[holo] (unchanged)
![unshifted_holo](https://github.com/Helium314/HeliBoard/assets/151087174/75b81611-cc55-4027-a7a0-f2729eded68a)


**shifted state:**
[material]
![material_shifted](https://github.com/Helium314/HeliBoard/assets/151087174/291b084b-5562-43fb-88b0-45113a3512f8)

[holo] (unchaged)
![unshifted_holo](https://github.com/Helium314/HeliBoard/assets/151087174/0c484574-abef-475c-8e05-96ab77c15d6a)


**caps lock:**
[material]
![material_locked](https://github.com/Helium314/HeliBoard/assets/151087174/88bab480-7276-446f-9a71-a09cb40cb7bd)

[holo]
![holo_locked](https://github.com/Helium314/HeliBoard/assets/151087174/942b6747-1fbb-4990-a672-e66319e3f6d9)

Credit to Google/Material Icons for the base icons, and @Uranusek for the ideas for the icons
Rounded theme has the same icons as the material theme, only that the icons are slightly rounded
(fixes #683)

Uranusek commented:
In my opinion, this icon does not fit other themes than Holo. What do you think about something like this? 

no shift - regular icon
pressed shift - add accent
caps lock - add a second arrow above the current one

codokie commented:
@Uranusek Thanks for chiming in, I agree that the holo may not be the best fit. So I adopted your idea for the light/rounded themes, you can see the result in the main post above.
Currently the shift and caps lock use the same icon so it would be less hassle to keep it that way and differentiate them with an accent only.
I personally think that adding accent only when there is a caps lock would look better most of the time if the accent color of the theme does not compliment very well the shift key icon

Uranusek commented:
Looks good. My idea was to change the icon only after pressing Caps Lock + adding an accent. Pressing Shift would only add an accent, not change the icon. I like the accent when I press the Shift key.

codokie commented:
as i said that is feasible but it's a personal preference most likely. 
If I am not mistaken the key for shifted state is currently shared with the caps lock so we would need to create a new key which would require to change many functions. Other keyboards like Gboard for example only change the keys icon.
I can easily switch between shifted and caps lock so that shifted would have accented color while caps lock will have regular color. Both will still have the same icon, I personally dislike the accented color for shifted state though

codokie commented:
@eliottmusk  I've also added a vibration so that it would be easier to tell we are in caps lock mode, should help when long pressing the shift key

Uranusek commented:
I understand. In Gboard, the icon changed when I pressed Caps Lock. I thought it would be easy to do the same in HeliBoard.

codokie commented:
@Uranusek  actually i was wrong it is very easy to change the icon of the caps lock.
currently we have 2 icons, I think 3 different icons for the 3 different shift states would be ideal, any suggestions what they should be?

Uranusek commented:
For now, I don't have a good idea for the third icon. Maybe add a second ^ above the current one 🤔 

I think two icons would also be sufficient. Just like in Gboard. For most, I think it would be intuitive.

codokie commented:
I've made a slight change, here is how it looks for the Holo theme with Brown accent:
unshifted (old)
![unshifted](https://github.com/Helium314/HeliBoard/assets/151087174/2c18a624-2412-43f9-8cfc-d6a291f93399)
shifted (old)
![shifted](https://github.com/Helium314/HeliBoard/assets/151087174/93b23829-ba1f-4baa-b5a0-772623771b2b)
caps lock (new) 
![caps](https://github.com/Helium314/HeliBoard/assets/151087174/355782dd-0794-4a5c-9550-0b06be73dc01)
From what i've seen Gboard does have 3 icons with a similar pattern to this.
I would like the light & rounded themes to match this pattern, but it's a bit difficult with current icon

codokie commented:
I made some changes to the icons (see the main post). Would really like feedback on whether it looks cool

BlackyHawky commented:
Perhaps you could use a simple arrow for Caps Lock mode while keeping the bottom bar?
In any case, the double arrow for shifted state is a good idea.

Quick question: will these icons be rounded for the Rounded style?

Uranusek commented:
For Holo it's great. For the rest of the themes I would do this:

unshifted: 
![322207478-23b19853-f4de-4a56-bffc-b72adad288e4](https://github.com/Helium314/HeliBoard/assets/135356908/dcc7bb85-1c96-409e-91fb-8ec2be95ed7b) 

shifted (with accent):
![322207478-23b19853-f4de-4a56-bffc-b72adad288e4](https://github.com/Helium314/HeliBoard/assets/135356908/920a0f1e-86f8-4fe3-a3a0-c6b0f32b4457)

caps lock:
![322260062-7d393cd2-6e5e-4323-ab3b-a13ecf3b1b0c](https://github.com/Helium314/HeliBoard/assets/135356908/79409091-a2cc-4e65-8a15-c8ce98f322c1)

codokie commented:
Thank you both for the feedback!

@Uranusek so would you rather have only 2 different icons? that may make it less obvious when shifted state is on, depending how the accented color suits the theme.

The following icon is actually called "caps lock". @BlackyHawky  I tend to agree with you that it would be a better match for shift locked state.
![image](https://github.com/Helium314/HeliBoard/assets/151087174/753766fe-f63f-470c-a3f3-41c6ab322f84)

codokie commented:
here is a quick demo of the new shift keys in Rounded/Brown theme:

https://github.com/Helium314/HeliBoard/assets/151087174/9d1cc53e-f69c-4c5f-afa9-0b61bfd18a34

Uranusek commented:
Yes. I think two icons would be more intuitive. I'm a little confused when I see three icons when switching. Moreover, each icon has become different. In my opinion, this is unnecessary complication :/ It was enough to add an icon for caps lock and leave the rest as they were.

codokie commented:
Upon further testing the added vibration does not really seem to be noticeable so I'm reverting that commit.
On a second thought the previous icon seem to make more sense for the caps lock because it matches the pattern of the holo theme
@Helium314 eventually it is up to you to decide if 3 icons are warranted, I personally think they are needed for distinguishing the different states

Uranusek commented:
I think this icon might be too small and not look good.
![322260202-213618b4-468b-4dc5-9bd5-a0ea2afecd9b](https://github.com/Helium314/HeliBoard/assets/135356908/174cdfda-3e2a-4e6a-85af-4128e4067998)

It would be good to record what these icons look like on the keyboard. Both the current ones and those from my idea. It will be easier to decide which ones look better.

codokie commented:
@Uranusek 
here is the new one (theme is material/blue gray)

https://github.com/Helium314/HeliBoard/assets/151087174/66a8effc-b546-457b-851e-e6d751428e33

and here is the old one

https://github.com/Helium314/HeliBoard/assets/151087174/377a660a-744e-4e88-ba27-927799edfc98

in my opinion the difference in size is not too big

Uranusek commented:
Looks nice. I wonder what the keyboard would look like with the changes I proposed 🤔

codokie commented:
here you go

https://github.com/Helium314/HeliBoard/assets/151087174/2cf2256b-fe9c-45e4-8dd2-bc018b8a5cd9

I think it is also a viable option, but in certain color/them combos the accent is subtle and it may be hard to tell if it shift is activated just by looking at the key

Uranusek commented:
Thanks. I like my option a bit more, but both are cool. 
Helium said it would be possible to customize the color of this icon in the future. Everyone will be able to choose their own, so the accent will be clearly visible. 
https://github.com/Helium314/HeliBoard/issues/529

BabyOilJohnson commented:
Hi, @codokie can u try my idea.
Just like the ideas of Uranusek, instead of the addition arrow replace it with another dash above the arrow.

Helium314 commented:
Finally some basic review:
Code changes are fine, and so is using 3 icons.

But I have to say I'm not happy with the icons, mainly I just don't like the double arrows.
My suggestion would be using the single `^` for unshifted state, and 
* either the current shift icon for shifted state (but normal color), and set accent color when locked
  * I ike this more
* or the  `^` icon also for shifted state (accent color), and the current icon (with accent color) when locked
  * this is more consistent with when the bar is added below the icon

(Holo looks fine)

codokie commented:
@Helium314  thanks for the feedback!
I prefer the second option you mentioned because [this icon](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Akeyboard_capslock%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4024) is called "Keyboard Capslock".

I modified the caps lock icon for the holo theme to be more like the regular shifted icon, just with a bar below the it.

I updated the main post so you can see how it looks

Helium314 commented:
> @Helium314 thanks for the feedback! I prefer the second option you mentioned because [this icon](https://fonts.google.com/icons?selected=Material%20Symbols%20Outlined%3Akeyboard_capslock%3AFILL%400%3Bwght%40400%3BGRAD%400%3Bopsz%4024) is called "Keyboard Capslock".

Using Android defaults is certainly a good reason to do it like that, thank for noticing this.

> I updated the main post so you can see how it looks

Thanks, though you missed the holo shift icon (it's fine in app, I already checked)

codokie commented:
> you missed the holo shift icon

My bad, must have uploaded the same image twice..

Glad you liked the new icons, thanks for taking the time to review and merge this PR
